p2p/trust/trustmetric.go

```diff
@@ -0,0 +1,545 @@
+		peerMetrics: make(map[string]*TrustMetric),
+		db:          db,
+		config:      tmc,
+	}
+
+	tms.BaseService = *cmn.NewBaseService(nil, "TrustMetricStore", tms)
+	return tms
+}
+
+// OnStart implements Service
+func (tms *TrustMetricStore) OnStart() error {
+	tms.BaseService.OnStart()
+
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	tms.loadFromDB()
+	return nil
+}
+
+// OnStop implements Service
+func (tms *TrustMetricStore) OnStop() {
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	// Stop all trust metric goroutines
+	for _, tm := range tms.peerMetrics {
+		tm.Stop()
+	}
+
+	tms.saveToDB()
+	tms.BaseService.OnStop()
+}
+
+// Size returns the number of entries in the trust metric store
+func (tms *TrustMetricStore) Size() int {
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	return tms.size()
+}
+
+// GetPeerTrustMetric returns a trust metric by peer key
+func (tms *TrustMetricStore) GetPeerTrustMetric(key string) *TrustMetric {
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	tm, ok := tms.peerMetrics[key]
+	if !ok {
+		// If the metric is not available, we will create it
+		tm = NewMetricWithConfig(tms.config)
+		if tm != nil {
+			// The metric needs to be in the map
+			tms.peerMetrics[key] = tm
+		}
+	}
+	return tm
+}
+
+// PeerDisconnected pauses the trust metric associated with the peer identified by the key
+func (tms *TrustMetricStore) PeerDisconnected(key string) {
+	tms.mtx.Lock()
+	defer tms.mtx.Unlock()
+
+	// If the Peer that disconnected has a metric, pause it
+	if tm, ok := tms.peerMetrics[key]; ok {
+		tm.Pause()
+	}
+}
+
+/* Private methods */
+
+// size returns the number of entries in the store without acquiring the mutex
+func (tms *TrustMetricStore) size() int {
+	return len(tms.peerMetrics)
+}
+
+/* Loading & Saving */
+/* Both of these methods assume the mutex has been acquired, since they write to the map */
+
+var trustMetricKey = []byte("trustMetricStore")
+
+type peerHistoryJSON struct {
+	NumIntervals int       `json:"intervals"`
+	History      []float64 `json:"history"`
+}
+
+// Loads the history data for the Peer identified by key from the store DB.
+// cmn.Panics if file is corrupt
+func (tms *TrustMetricStore) loadFromDB() bool {
+	// Obtain the history data we have so far
+	bytes := tms.db.Get(trustMetricKey)
+	if bytes == nil {
+		return false
+	}
+
+	peers := make(map[string]peerHistoryJSON, 0)
+	err := json.Unmarshal(bytes, &peers)
+	if err != nil {
+		cmn.PanicCrisis(cmn.Fmt("Could not unmarchal Trust Metric Store DB data: %v", err))
+	}
+
+	// If history data exists in the file,
+	// load it into trust metrics and recalc
+	for key, p := range peers {
+		tm := NewMetricWithConfig(tms.config)
+
+		// Restore the number of time intervals we have previously tracked
+		if p.NumIntervals > tm.maxIntervals {
+			p.NumIntervals = tm.maxIntervals
+		}
+		tm.numIntervals = p.NumIntervals
+		// Restore the history and its current size
+		if len(p.History) > tm.historyMaxSize {
+			p.History = p.History[:tm.historyMaxSize]
```

Comment:
so `History` is most recent first. This means we have to use `append` in a "backwards" way (ie. `append([]xxx{x}, xxx...)` vs. `append(xxx, x)`). Is there a particular reason to do it this way rather than put the most recent elements on the end? Seems like this would be less performant since if I'm not mistaken we'll be reallocating memory every time.

Reply:
okay. History is now most recent last. I have updated the calculations to handle this properly and reduced memory allocations related to the history data. Running some tests right now